File: src/n_generics/generics.java
// package n_generics

import java.util.Collections;
import java.util.List;

void main() {
    final var box = new Box<Integer>(1);
    final var box2 = new Box<String>("Hello");
    final var box3 = new Box<>(1);

    System.out.println(box.get());
    System.out.println(box2.get());
    System.out.println(box3.get());

    List<String> list = singletonList("Hello");
    System.out.println(list);

    String basicToString = basicToString(1);
    System.out.println(basicToString);
}

public static class Box<T> {
    private final T t;

    public Box(T t) {
        this.t = t;
    }

    public T get() {
        return t;
    }
}

public static <T> List<T> singletonList(T item) {
    return Collections.singletonList(item);
}

public static <T> String basicToString(T item) {
    return STR."Value: \{item}";
}

File: src/n_generics/generics.kt
package n_generics

fun main(args: Array<String>) {
    val box: Box<Int> = Box(1)
    val box2: Box<String> = Box("Hello")
    val box3 = Box(1) // 1 has type Int, so the compiler figures out that it is Box<Int>

    println(box.value)
    println(box2.value)
    println(box3.value)

    val list = singletonList("Hello")
    println(list)

    val basicToString = 1.basicToString()
    println(basicToString)
}

class Box<T>(t: T) {
    val value = t
}


fun <T> singletonList(item: T): List<T> {
    return listOf(item)
}

fun <T> T.basicToString(): String {
    return "Value: $this"
}